bsdtty.1
.Dd August 13, 2018
.Dt BSDTTY 1
.Sh NAME
.Nm bsdtty
.Nd BSD RTTY Client
.Sh SYNOPSIS
.Nm
.Op Fl ahT
.Op Fl c charset
.Op Fl C callsign
.Op Fl d baud_denominator
.Op Fl f freq_offset
.Op Fl i rigctld_host
.Op Fl I rigctld_port
.Op Fl l logfile
.Op Fl m mark_freq
.Op Fl n baud_numerator
.Op Fl N suppress
.Op Fl p dsp_device
.Op Fl P xmlrpc_port
.Op Fl q bp_filter_q
.Op Fl Q lp_filter_q
.Op Fl r dsp_rate
.Op Fl s space_freq
.Op Fl t tty_device
.Op Fl x xmlrpc_host
.Op Fl 1 f1_macro
.Op Fl 2 f2_macro
.Op Fl 3 f3_macro
.Op Fl 4 f4_macro
.Op Fl 5 f5_macro
.Op Fl 6 f6_macro
.Op Fl 7 f7_macro
.Op Fl 8 f8_macro
.Op Fl 9 f9_macro
.Op Fl 0 f10_macro
.Sh DESCRIPTION
The
.Nm
utility provides an RTTY terminal with a "crossed bananas" and waterfall display
on an ncurses terminal.
.Pp
The options are as follows:
.Bl -tag -width indent
.It Fl a
Enable AFSK mode.
.It Fl c Ar charset
Use the specified charset.
0 indicates the most common ITA2 variant, 1 indicates the US-TTY variant,
and 2 indicates a strict ITA2 with no national characters.
Default is 0.
.It Fl C Ar callsign
Specifies the callsign of the station operator.
Default is W8BSD.
.It Fl d Ar baud_denominator
The deominator of the baudrate to use.
Default is 1000
.It Fl f Ar freq_offset
Frequency offset from the VFO value to the mark frequency.
This is only used if rigctld support is enabled.
Default is 170.
.It Fl h
Displays usage help.
.It Fl i Ar rigctld_host
Hostname to connect to for rigctld rig control.
If hostname is a zero-length string, rigctld support is disabled.
.It Fl I Ar rigctld_port
Port number to connect to for rigctld rig control.
If port is zero, rigctld support is disabled.
.It Fl l Ar logfile
Specifies the path and filename of the logfile for TX and RX adata.
Default is bsdtty.log
.It Fl m Ar mark_freq
The mark frequency in the receive and transmit audio.
Default is 2125.
.It Fl N Ar suppress
By default, bsdtty adds a CRLF to the start of all transmissions, and
a single space to the end.  If
.Ar suppress
contains an s, the start CRLF
is suppressed.  If
.Ar suppress
contains an e, the ending space is suppressed.
.It Fl p Ar dsp_device
Specifies the full path to the DSP device to use.
Default is /dev/dsp8
.It Fl P Ar xmlrpc_port
Sets the XML-RPC port for Fldigi emulation.
Default is 7362
.It Fl q Ar bp_filter_q
If
.Nm
was compiled without matched filters, specifies the Q of the mark/space bandpass
filters.
Default is 10.
.It Fl Q Ar lp_filter_q
Specifies the Q of the lowpass filter used for envelope detection.
Default is 0.5
.It Fl r Ar dsp_rate
Sets the rate for recording and playback via the DSP.
Default is 8000.
.It Fl s Ar space_freq
The space frequency in the receive and transmit audio.
Default is 2295.
.It Fl t Ar tty_device
Full pathname to tty device for FSK and RTS PTT.
Default is /dev/ttyu9.
.It Fl x Ar xmlrpc_host
Hostname to listen on for XML-RPC requests (Fldigi emulation).
Default is an empty string.
.It Fl 1 2 3 4 5 6 7 8 9 0
Specifies the macros to send when Fx is pressed.
.Fl 0
specifies the F10 macro.
.El
.Pp
When
.Nm
starts, it is in RX mode.  TX mode can be entered by typing any valid
baudot characters, or via the TAB key.
Transmission will continue until TAB is pressed again.
.Pp
When in RX mode, the following extra commands can be used:
.Bl -tag -width indent
.It `
Toggles reverse mode.  In AFSK mode (when
.Fl a
is specified), the TX and RX signals are reversed.
In FSK mode (the default), on the RX signals are reverse.
.It [
Switches to the previous character set.
This only affects characters transmitted and received after the change is made.
.It ]
Switches to the next character set.
This only affects characters transmitted and received after the change is made.
.It BACKSPACE
Opens the configuration editor.
In the configuration editor, arrow keys move between and inside fields.
ENTER will save changes, and CTRL-C aborts editing.
.It CTRL-C
Exits bsdtty.
.It CTRL-L
Clears the RX window.
.It CTRL-W
Toggles the tuning aid display. Cycles between the crossed banana, waterfall, and none
tuning aids.
.It LEFT
Lowers the character squelch.  This specifies the number of characters that must be
received one after another before any characters are displayed.
.It RIGHT
Increases the character squelch.  This specifies the number of characters that must be
received one after another before any characters are displayed.
.It UP
Increments the current serial number.
.It DOWN
Decrements the current serial number.
.It LEFT-CLICK
Selects a callsign.  The callsign is displayed in the status line, and can be send via
the ` character in a macro.
.It RIGHT-CLICK
Toggles the shift in the word under the mouse.
The toggle extends in both directions to the next whitespace or shift.
.El
.Sh MACROS
.Pp
Macros are sent using the functions keys F1 to F10.
.Pp
Sending a macro which starts with "CQ CQ" or ends with " CQ" will clear the RX window.
.Pp
Special characters can be used in macros.
.Bl -tag -width indent
.It \\
Sends the value specified in the Callsign configuration option of via the
.Fl C
argument.
.It `
Sends the currently captured callsign (via LEFT-CLICK).
.It [
Sends a CRLF pair.
.It ~
Disables TX and terminates the macro.
.It _
Sends a space character, terminates the macro, but leaves
.Nm
in transmit mode.
.It ^
Increments the serial number then sends it as at least three
digits.
.It %
Sends the serial number as at least three digits.
.El
.Pp
Some default macros are pre-configured in
.Nm .
.Bl -tag -width indent
.It F2
"CQ CQ CQ CQ CQ CQ DE \\ \\ \\ PSE K~"
.It F3
"\ "
.It F4
"` DE `~"
.El
.Sh THE SCREEN
The screen is divided into three sections sub-windows
.Bl -tag -width indent
.It Status
The status window is one line high and contains the following:
.Bl -tag -width indent
.It REV
If REV is displayed, indicates that reverse mode is enabled.
In AFSK mode, this indicates that both RX and TX tones are reversed.
In FSK mode, indicates that just RX tones are reversed.
.It Ar charset
The abbreviation of the current character set:
.Bl -tag -width indent
.It ITA2
The most common ITA2 character set.  This includes the '#' figure (shifted D),
the '@' figure (shifted F), the '$' figure (shifted H), and the '*' figure (shifted G).
.It USTTY
The most common US variant of baudot.  This swaps the BEL and ' characters, includes
the '$' figure (shifted D), the '!' figure (shifted F), the '"' figure (shifted Z),
the '#' figure (shifted H), the '&' figure (shifted G), and the ';' figure (shifted V)
Is missing the '+' and '=' figures from the standard ITA2 character set (replaced by
'&' and ';').
.It ITA2(S)
This is the official ITA2 charset from ITU-T S.1.  It is identical to 
.Ar ITA2
except shifted D is the ENQ character, and shifted F, H, and G ('@', '$', and '*') are not valid.
.El
.It Ar frequency
If rigctld is enabled, displays the current mark frequency.
The VFO offset should be set for this to be accurate.
.Pp
The background colour of the frequency changes based on the current frequency in
the HF bands.  If the background is green, the current frequency is in the normal RTTY
subbands.  If the background is black, the frequency is in the subbands normally used
in contests.  If the background is yellow or brown, indicates that RTTY is legal in the US
at this frequency, but it is almost never used for RTTY.  If the background is red, indicates
that RTTY is not allowed on the current frequency.
.Pp
The background will only be red if the configured callsign is in the US.  It may also be incorrect
for US hams in ITU region 1.
.It Ar mode
Indicates the mode the rg is currently in (ie: 'USB', 'LSB', 'RTTY', etc).
.It Ar callsign
The current callsign selected by left-click and used for the ` macro character.
.It SQL Ar x
Where x is a value between 1 and 9 inclusive.
This indicates the number of characters that must be received after each other before
any characters are displayed.
This is an experimental feature, and likely shouldn't be used.
.It Ar serial
The current serial number formatted as at least three digits.
.It Ar VU
Ad the end of the status window is a VU meder.
.El
.It RX
Shows decoded characters.
It is in this window that the left and right mouse buttons have effect.
.It TX
The TX window shows what you have send and are currently sending.
In RX mode, can be toggled with the CTRL-W key to one of the two tuning aids.
.It Tuning Aid
Displays either a crossed bananas graph or a waterfall in RX mode.
Toggle this using CTRL-W.
.El


bsdtty.c
/*-
 * Copyright (c) 2018 Stephen Hurd, W8BSD
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 *
 */

#include <sys/ioctl.h>

#include <ctype.h>
#include <errno.h>
#include <fcntl.h>
#include <inttypes.h>
#include <limits.h>
#include <locale.h>
#include <netdb.h>
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>
#include <stdnoreturn.h>
#include <string.h>
#include <termios.h>
#include <time.h>
#include <unistd.h>

#include "afsk_send.h"
#include "fsk_send.h"
#include "baudot.h"
#include "bsdtty.h"
#include "fldigi_xmlrpc.h"
#include "fsk_demod.h"
#include "rigctl.h"
#include "ui.h"

static bool do_tx(int *rxstate);
static void done(void);
static void handle_rx_char(char ch);
static void input_loop(void);
static void send_char(const char ch);
static void send_rtty_char(char ch);
static void setup_log(void);
noreturn static void usage(const char *cmd);
static void setup_defaults(void);
static void fix_config(void);
struct send_fsk_api *send_fsk;

struct bt_settings settings = {
	.baud_numerator = 1000,
	.baud_denominator = 22,
	.dsp_rate = 8000,
	.bp_filter_q = 10,
	.lp_filter_q = 0.5,
	.mark_freq = 2125,
	.space_freq = 2295,
	.charset = 0,
	.afsk = false,
	.ctl_ptt = false,
	.freq_offset = 170,
	.rigctld_port = 4532,
	.xmlrpc_port = 7362
};

/* RX in reverse mode */
bool reverse = false;

/* Log thing */
static FILE *log_file;

static char sync_buffer[9];
static int sb_chars;
static int sync_squelch = 1;
static bool rxfigs;
static bool txfigs;
static bool send_start_crlf = true;
static bool send_end_space = true;

char *their_callsign;
unsigned serial;

int main(int argc, char **argv)
{
	char *c;
	int ch;

	load_config();

	while ((ch = getopt(argc, argv, "ac:C:d:f:hl:i:I:m:n:N:p:P:q:Q:r:s:t:T1:x:2:3:4:5:6:7:8:9:0:")) != -1) {
		while (optarg && isspace(*optarg))
			optarg++;
		switch (ch) {
			case '0':
				settings.macros[9] = strdup(optarg);
				break;
			case '1':
			case '2':
			case '3':
			case '4':
			case '5':
			case '6':
			case '7':
			case '8':
			case '9':
				settings.macros[ch-'1'] = strdup(optarg);
				break;
			case 'a':
				settings.afsk = true;
				break;
			case 'c':
				settings.charset = strtoi(optarg, NULL, 10);
				if (settings.charset < 0 || settings.charset >= charset_count)
					settings.charset = 0;
				break;
			case 'C':
				settings.callsign = strdup(optarg);
				break;
			case 'd':	// baud_denominator
				settings.baud_denominator = strtoi(optarg, NULL, 10);
				break;
			case 'f':
				settings.freq_offset = strtoi(optarg, NULL, 10);
				break;
			case 'F':
				settings.freq_offset = strtoi(optarg, NULL, 10);
				break;
			case 'h':
				usage(argv[0]);
			case 'i':
				settings.rigctld_host = strdup(optarg);
				break;
			case 'I':
				settings.rigctld_port = strtoi(optarg, NULL, 10);
				break;
			case 'l':	// log_name
				settings.log_name = strdup(optarg);
				break;
			case 'm':	// mark_freq
				settings.mark_freq = strtod(optarg, NULL);
				break;
			case 'n':	// baud_numerator
				settings.baud_numerator = strtoi(optarg, NULL, 10);
				break;
			case 'N':
				for (c = optarg; *c; c++) {
					switch (*c) {
						case 's':
							send_start_crlf = false;
							break;
						case 'e':
							send_end_space = false;
							break;
					}
				}
				break;
			case 'p':	// dsp_name
				settings.dsp_name = strdup(optarg);
				break;
			case 'P':
				settings.xmlrpc_port = strtoi(optarg, NULL, 10);
				break;
			case 'q':	// bp_filter_q
				settings.bp_filter_q = strtod(optarg, NULL);
				break;
			case 'Q':	// lp_filter_q1
				settings.lp_filter_q = strtod(optarg, NULL);
				break;
			case 'r':	// dsp_rate
				settings.dsp_rate = strtoi(optarg, NULL, 10);
				break;
			case 's':	// space_freq
				settings.space_freq = strtod(optarg, NULL);
				break;
			case 't':	// tty_name
				settings.tty_name = strdup(optarg);
				break;
			case 'T':
				settings.ctl_ptt = true;
				break;
			case 'x':
				settings.xmlrpc_host = strdup(optarg);
				break;
			default:
				usage(argv[0]);
		}
	}

	setup_defaults();
	fix_config();

	setlocale(LC_ALL, "");
	atexit(done);

	/*
	 * We want to set up the sockets before we setup curses since
	 * it can spin writing to stderr.
	 */
	setup_xmlrpc();

	// Now set up curses
	setup_curses();

	// Set up the FSK stuff.
	setup_rx();

	// And AFSK
	if (settings.afsk)
		send_fsk = &afsk_api;
	else
		send_fsk = &fsk_api;

	send_fsk->setup();

	// Set up the log file
	setup_log();

	display_charset(charset_name(settings.charset));
	update_squelch(sync_squelch);
	update_serial(serial);

	setup_rig_control();

	// Finally, do the thing.
	input_loop();
	return EXIT_SUCCESS;
}

void
reinit(void)
{
	load_config();
	fix_config();

	// Set up the FSK stuff.
	setup_rx();
	if (settings.afsk)
		send_fsk = &afsk_api;
	else
		send_fsk = &fsk_api;
	send_fsk->setup();

	// Set up the log file
	setup_log();

	setup_rig_control();
}

static void
setup_defaults(void)
{
	/*
	 * We do this here so NULL means uninitialized and we can always
	 * free() strings.
	 */

	if (settings.log_name == NULL)
		settings.log_name = strdup("bsdtty.log");
	if (settings.tty_name == NULL)
		settings.tty_name = strdup("/dev/ttyu9");
	if (settings.dsp_name == NULL)
		settings.dsp_name = strdup("/dev/dsp8");
	if (settings.macros[1] == NULL)
		settings.macros[1] = strdup("CQ CQ CQ CQ CQ CQ DE \\ \\ \\ PSE K\t");
	if (settings.macros[2] == NULL)
		settings.macros[2] = strdup("\\ ");
	if (settings.macros[3] == NULL)
		settings.macros[3] = strdup("` DE \\\t");
	if (settings.callsign == NULL)
		settings.callsign = strdup("W8BSD");
	if (settings.rigctld_host == NULL)
		settings.rigctld_host = strdup("localhost");
}

static void
setup_log(void)
{
	time_t now;

	if (log_file != NULL)
		fclose(log_file);
	if (log_file)
		fclose(log_file);
	log_file = fopen(settings.log_name, "a");
	if (log_file == NULL)
		printf_errno("opening log file");
	now = time(NULL);
	if (log_file != NULL)
		fprintf(log_file, "\n\n%s\n", ctime(&now));
}

static void
handle_rx_char(char ch)
{
	switch(ch) {
		case 0:
			return;
		case 0x07:	// BEL
			return;
		case 0x0f:	// LTRS
			rxfigs = false;
			return;
		case 0x0e:	// FIGS
			rxfigs = true;
			return;
		case ' ':
			rxfigs = false;	// USOS
	}
	if (log_file != NULL)
		fwrite(&ch, 1, 1, log_file);
	fldigi_add_rx(ch);
	write_rx(ch);
}

static void
input_loop(void)
{
	int rxstate = -1;
	int i;
	char ch;

	while (1) {
		handle_xmlrpc();
		if (get_rig_ptt()) {	// TX Mode
			if (!do_tx(&rxstate))
				return;
			rxstate = -1;
		}
		else {
			if (check_input()) {
				if (!do_tx(&rxstate))
					return;
				continue;
			}

			rxstate = get_rtty_ch(rxstate);
			if (rxstate < 0) {
				sb_chars = 0;
				rxfigs = false;
				continue;
			}
			if (rxstate > 0x20)
				printf_errno("got a==%d", rxstate);
			ch = baudot2asc(rxstate, rxfigs);
			if (sb_chars < sync_squelch) {
				sync_buffer[sb_chars++] = ch;
				if (sb_chars == sync_squelch) {
					for (i = 0; i < sync_squelch; i++)
						handle_rx_char(sync_buffer[i]);
					continue;
				}
				else
					continue;
			}
			handle_rx_char(ch);
		}
	}
}

static bool
do_tx(int *rxstate)
{
	int ch;

	ch = get_input();
	switch (ch) {
		case -1:
			send_fsk->diddle();
			break;
		case 3:
			return false;
		case RTTY_FKEY(1):
			do_macro(1);
			break;
		case RTTY_FKEY(2):
			do_macro(2);
			break;
		case RTTY_FKEY(3):
			do_macro(3);
			break;
		case RTTY_FKEY(4):
			do_macro(4);
			break;
		case RTTY_FKEY(5):
			do_macro(5);
			break;
		case RTTY_FKEY(6):
			do_macro(6);
			break;
		case RTTY_FKEY(7):
			do_macro(7);
			break;
		case RTTY_FKEY(8):
			do_macro(8);
			break;
		case RTTY_FKEY(9):
			do_macro(9);
			break;
		case RTTY_FKEY(10):
			do_macro(10);
			break;
		case RTTY_KEY_LEFT:
			sync_squelch--;
			if (sync_squelch < 1)
				sync_squelch = 1;
			update_squelch(sync_squelch);
			break;
		case RTTY_KEY_RIGHT:
			sync_squelch++;
			if (sync_squelch > 9)
				sync_squelch = 9;
			update_squelch(sync_squelch);
			break;
		case RTTY_KEY_UP:
			serial++;
			update_serial(serial);
			break;
		case RTTY_KEY_DOWN:
			if (serial)
				serial--;
			update_serial(serial);
			break;
		case RTTY_KEY_REFRESH:
			display_charset(charset_name(settings.charset));
			update_squelch(sync_squelch);
			show_reverse(reverse);
			update_captured_call(their_callsign);
			update_serial(serial);
			break;
		case '`':
			toggle_reverse(&reverse);
			send_fsk->toggle_reverse();
			*rxstate = -1;
			break;
		case '[':
			settings.charset--;
			if (settings.charset < 0)
				settings.charset = charset_count - 1;
			display_charset(charset_name(settings.charset));
			break;
		case ']':
			settings.charset++;
			if (settings.charset == charset_count)
				settings.charset = 0;
			display_charset(charset_name(settings.charset));
			break;
		case '\\':
			reset_tuning_aid();
			break;
		case 23:
			if (!get_rig_ptt())
				toggle_tuning_aid();
			break;
		case 0x7f:
		case 0x08:
			if (!get_rig_ptt()) {
				change_settings();
				*rxstate = -1;
			}
			break;
		default:
			send_char(ch);
			break;
	}
	return true;
}

void
send_string(const char *str)
{
	if (str == NULL)
		return;

	for (; *str; str++)
		send_char(*str);
}

bool
do_macro(int fkey)
{
	size_t len;
	size_t i;
	int m;
	char buf[11];

	m = fkey - 1;
	if (settings.macros[m] == NULL)
		return true;
	if (strncasecmp(settings.macros[m], "CQ CQ", 5) == 0)
		clear_rx_window();
	len = strlen(settings.macros[m]);
	for (i = 0; i < len; i++) {
		switch (settings.macros[m][i]) {
			case '\\':
				send_string(settings.callsign);
				break;
			case '`':
				send_string(their_callsign);
				break;
			case '[':
				send_char('\r');
				break;
			case ']':
				// TODO: CR is expanded to CRLF...
				send_char('\n');
				break;
			case '^':
				serial++;
				/* Fall-through */
			case '%':
				sprintf(buf, "%d", serial);
				send_string(buf);
				update_serial(serial);
				break;
			default:
				send_char(settings.macros[m][i]);
				break;
		}
	}
	if (len > 2)
		if (strcasecmp(settings.macros[m] + len - 3, " CQ") == 0)
			clear_rx_window();
	return true;
}

static void
send_char(const char ch)
{
	const char fstr[] = "\x1f\x1b"; // LTRS, FIGS
	char bch;
	bool rts;
	char ach;
	time_t now;
	static uint64_t freq = 0;
	static char mode[16];

	bch = asc2baudot(ch, txfigs);
	rts = get_rig_ptt();

	if (ch == '\t' || (!rts && bch != 0)) {
		rts = !rts;
		if (!rts) {
			if (send_end_space)
				send_rtty_char(4);
			send_fsk->end_tx();
		}
		if (rts) {
			get_rig_freq_mode(&freq, mode, sizeof(mode));
			if (freq)
				freq += settings.freq_offset;
		}
		now = time(NULL);
		if (log_file != NULL) {
			if (freq == 0)
				fprintf(log_file, "\n------- %s of transmission (%.24s) -------\n", rts ? "Start" : "End", ctime(&now));
			else
				fprintf(log_file, "\n------- %s of transmission (%.24s) on %s (%s) -------\n", rts ? "Start" : "End", ctime(&now), format_freq(freq), mode);
			fflush(log_file);
		}
		mark_tx_extent(rts);
		set_rig_ptt(rts);
		if (rts) {
			/* 
			 * Start with a byte length of mark to help sync...
			 * This also covers the RX -> TX switching time
			 * due to the relay.
			 */
			send_fsk->send_preamble();
			txfigs = false;
			/*
			 * Per ITU-T S.1, the FIRST symbol should be a
			 * shift... since it's most likely to be lost,
			 * a LTRS is the safest, since a FIGS will get
			 * repeated after the CRLF.
			 */
			send_rtty_char(0x1f);
			if (send_start_crlf) {
				send_rtty_char(8);
				send_rtty_char(2);
			}
		}
	}
	if (bch == 0)
		return;
	if (bch) {
		// Send FIGS/LTRS as needed
		if ((!!(bch & 0x20)) != txfigs) {
			txfigs = !!(bch & 0x20);
			send_rtty_char(fstr[txfigs]);
		}
		/* We do this to ensure it's valid baudot */
		ach = baudot2asc(bch & 0x1f, bch & 0x20);
		switch (ach) {
			case 0:
				break;
			case 0x0e:
				txfigs = false;
				break;
			case 0x0f:
				txfigs = true;
				break;
			case '\r':
				write_tx('\r');
				if (log_file != NULL)
					fwrite("\r\n", 2, 1, log_file);
				break;
			default:
				write_tx(ach);
				if (log_file != NULL)
					fwrite(&ach, 1, 1, log_file);
				break;
		}
		if (ach == ' ')
			txfigs = false;	// USOS
		bch &= 0x1f;
		send_rtty_char(bch);
		if (bch == 0x08)
			send_rtty_char(2);
	}
}

static void
done(void)
{
	set_rig_ptt(false);
	close_sockets();
}

int
strtoi(const char *nptr, char **endptr, int base)
{
	int ret;
	long l;

	l = strtol(nptr, endptr, base);
	ret = (int)l;
	if (errno == ERANGE && (l == LONG_MIN || l == LONG_MAX)) {
		switch (l) {
			case LONG_MIN:
				ret = INT_MIN;
				break;
			case LONG_MAX:
				ret = INT_MAX;
				break;
		}
	}
	if (l < INT_MIN) {
		ret = INT_MIN;
		errno = ERANGE;
	}
	if (l > INT_MAX) {
		ret = INT_MAX;
		errno = ERANGE;
	}

	return ret;
}

unsigned int
strtoui(const char *nptr, char **endptr, int base)
{
	unsigned int ret;
	unsigned long l;

	l = strtoul(nptr, endptr, base);
	ret = (unsigned int)l;
	if (errno == ERANGE && l == ULONG_MAX)
		ret = UINT_MAX;
	if (l > UINT_MAX) {
		ret = UINT_MAX;
		errno = ERANGE;
	}

	return ret;
}

noreturn static void
usage(const char *cmd)
{
	printf("\n"
	       "Usage: %s [<arg>...]\n"
	       "\n"
	       "ARG Description                  Default\n"
	       "-t  TTY device name              /dev/ttyu9\n"
	       "-p  DSP device name              /dev/dsp8\n"
	       "-m  Mark audio frequency         2125.0\n"
	       "-s  Space audio frequency        2295.0\n"
	       "-n  Baudrate Numerator           1000\n"
	       "-d  Baudrate Denominator         22\n"
	       "-l  Logfile name                 bsdtty.log\n"
	       "-r  DSP rate                     16000\n"
	       "-q  Bandpass filter Q            10.0\n"
	       "-Q  Envelope lowpass filter Q    0.5\n"
	       "-1  F1 Macro                     <empty>\n"
	       "-2  F2 Macro                     \"CQ CQ CQ CQ CQ CQ DE W8BSD W8BSD W8BSD PSE K\"\n"
	       "-3  F3 Macro                     \"\\ \"\n"
	       "-4  F4 Macro                     \"` DE \\\"\n"
	       "-5  F5 Macro                     <empty>\n"
	       "-6  F6 Macro                     <empty>\n"
	       "-7  F7 Macro                     <empty>\n"
	       "-8  F8 Macro                     <empty>\n"
	       "-9  F9 Macro                     <empty>\n"
	       "-0  F10 Macro                    <empty>\n"
	       "-c  Charset to use               0\n"
	       "-a  Use AFSK (no argument)\n"
	       "-C  Callsign                     \"W8BSD\"\n"
	       "-T  Use rig control PTT (no argument)\n"
	       "-f  VFO frequency offset         170\n"
	       "-x  XML-RPC host name            \"localhost\"\n"
	       "-P  XML-RPC port                 7362\n"
	       "-N[s][e]\n"
	       "    Turns off automatically added text.  If 's' is included,\n"
	       "    the starting CRLF is not sent.  If 'e' is included, the\n"
	       "    ending space is not sent.\n"
	       "-i  rigctld host name            \"localhost\"\n"
	       "-I  rigctld port                 4532\n"
	       "\n", cmd);
	exit(EXIT_FAILURE);
}

static void
send_rtty_char(char ch)
{
	send_fsk->send_char(ch);
}

void
captured_callsign(const char *str)
{
	if (their_callsign) {
		free(their_callsign);
		their_callsign = NULL;
	}
	if (str == NULL || str[0] == 0)
		return;
	their_callsign = strdup(str);
}

static void
fix_config(void)
{
	if (settings.dsp_name == NULL)
		settings.dsp_name = strdup("/dev/dsp");
	if (settings.mark_freq < 1)
		settings.mark_freq = 2125;
	if (settings.mark_freq > settings.dsp_rate / 2)
		settings.mark_freq = settings.dsp_rate / 2 - 170;
	if (settings.space_freq < 1)
		settings.space_freq = 2295;
	if (settings.space_freq > settings.dsp_rate / 2)
		settings.space_freq = settings.dsp_rate / 2;
	if (settings.dsp_rate < 8000)
		settings.dsp_rate = 8000;
	if (settings.baud_denominator < 1)
		settings.baud_denominator = 1;
	if (settings.baud_numerator < 1)
		settings.baud_numerator = 1;
	if (settings.charset < 0)
		settings.charset = 0;
	if (settings.charset >= charset_count)
		settings.charset = 0;
	if (settings.rigctld_host == NULL || settings.rigctld_host[0] == 0 || settings.rigctld_port == 0)
		settings.ctl_ptt = false;
}

const char *
format_freq(uint64_t freq)
{
	static char fstr[32];
	const char *prefix = " kMGTPEZY";
	int pc = 0;
	int pos;
	char *ch;

	fstr[0] = 0;
	if (freq) {
		sprintf(fstr, "%" PRIu64, freq);
		for (pos = strlen(fstr) - 3; pos > 0; pos -= 3) {
			memmove(&fstr[pos]+1, &fstr[pos], strlen(&fstr[pos])+1);
			fstr[pos] = '.';
		}
	}
	while (strlen(fstr) > 11) {
		ch = strrchr(fstr, '.');
		if (ch == NULL)
			printf_errno("unable to find dot in freq \"%s\"", fstr);
		*ch = 0;
		pc++;
	}

	ch = strrchr(fstr, 0);
	if (ch == NULL)
		printf_errno("unable to find end of string");
	*(ch++) = prefix[pc];
	*(ch++) = 'H';
	*(ch++) = 'z';
	*ch = 0;

	return fstr;
}
